File: materialpreference/src/main/java/com/anggrayudi/materialpreference/widget/AspSwitchCompat.java
package com.anggrayudi.materialpreference.widget;

import android.content.Context;
import android.util.AttributeSet;

import androidx.annotation.RestrictTo;
import androidx.appcompat.widget.SwitchCompat;

/**
 * Works around https://code.google.com/p/android/issues/detail?id=196652.
 */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
public class AspSwitchCompat extends SwitchCompat {
    private boolean isInSetChecked = false;

    public AspSwitchCompat(Context context) {
        super(context);
    }

    public AspSwitchCompat(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public AspSwitchCompat(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    public void setChecked(boolean checked) {
        isInSetChecked = true;
        super.setChecked(checked);
        isInSetChecked = false;
    }

    @Override
    public boolean isShown() {
        if (isInSetChecked) {
            return getVisibility() == VISIBLE;
        }
        return super.isShown();
    }
}

File: materialpreference/src/main/java/com/anggrayudi/materialpreference/widget/AspSwitchCompat.kt
package com.anggrayudi.materialpreference.widget

import android.content.Context
import android.util.AttributeSet

import androidx.annotation.RestrictTo
import androidx.appcompat.widget.SwitchCompat

/**
 * Works around https://code.google.com/p/android/issues/detail?id=196652
 */
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
class AspSwitchCompat : SwitchCompat {
    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr)

    private var isInSetChecked = false

    override fun isShown(): Boolean = if (isInSetChecked) { visibility == VISIBLE } else super.isShown()

    override fun setChecked(checked: Boolean) {
        isInSetChecked = true
        super.setChecked(checked)
        isInSetChecked = false
    }
}
